### DISCUSSION TITLE
Augmenting Custom Sheet

### DISCUSSION THREAD

**Discussion by lightandshadow68** [pointfreeco/swift-composable-architecture#2071] [Beta]
What's the best way to extend a custom sheet modifier?

Since SwiftUI doesn't provide a way to specify a sheet using a FormSheet presentation, I'm using a custom sheet modifier that leverages UIKit under the hood to gain the same effect. I'd like to create augmented TCA navigation wrappers for this presentation, but the TCA navigation extensions use internal store methods to do much of the heavy lifting.  

What options do I have? Is there some special import I can use to make these available?

**Comment by mbrandonw** [comment]
Hey @lightandshadow68, unfortunately those tools are not in the navigation beta yet, but we do hope to improve the situation before the final release.

**Reply by grdsdev** [reply]
~~Hey @mbrandonw I'm trying to accomplish the same as @lightandshadow68, now that the navigation tools got released, is there any way of doing this? It seems that it still uses some internal methods.~~

Just saw that you can `@_spi(Presentation) import ComposableArchitecture` to get access to it.

**Reply by mbrandonw** [reply]
Ah yes, give it a whirl and let us know how it works out for you! We are eager to get feedback on it so that we can figure out what the final shape of the tools should be.

**Reply by lightandshadow68** [reply]
I've just found the SPI additions. The API looks nice!. However, I'm running into an issue when dismissing my custom sheet modifier. I have a few previews setup with vanilla SwiftUI, and the binding is set to nil. But I can't figure out how to get the corresponding state set to nil when dismissed with the Presentation additions. 

Pretty much duplicated the `.sheet` version and swapped them out with my `FormSheet` variant. 

```
@available(iOS 14, tvOS 14, watchOS 7, *)
  @available(macOS, unavailable)
  public func formSheet<State, Action, Content: View>(
    store: Store<PresentationState<State>, PresentationAction<Action>>,
    onDismiss: (() -> Void)? = nil,
    @ViewBuilder content: @escaping (Store<State, Action>) -> Content
  ) -> some View {
    self.presentation(store: store) { `self`, $item, destination in
      self.formSheet(item: $item, onDismiss: onDismiss) { _ in
        destination(content)
      }
    }
  }
```

Is there anything unusual I should be on the lookout for? It presents correctly and dismisses when the state is set to nil manually. But swiping down and calling `dismiss` on the environment method just doesn't result in clearing out the state.

**Reply by lightandshadow68** [reply]
Update: changing the call site to `.sheet` results in the presentation state being set to nil. So, it seems to be specific to my custom presentation, which uses UIKit to show content with the `.formSheet` presentation style. What's odd is that it seems to work fine with vanilla SwiftUI.

**Reply by lightandshadow68** [reply]
Tried wrapping a custom, third party custom sheet from SheeKit. The state is cleared when swiping down or clicking outside the sheet, but calling @Environment dismiss from inside the presented view doesn't clear the state. 

https://github.com/edudnyk/SheeKit

**Reply by lightandshadow68** [reply]
Ends up, there is a unique, dedicated @Environment variable for SheeKit presented views. Using that works just fine and the state is cleared.

**Reply by stephencelis** [reply]
Glad you figured it out! It definitely sounded like a view-layer issue if the binding wasn't being `nil`-ed out at that layer.
